### ISSUE TITLE
Warning: Received `true` for non-boolean attribute `gray`.

### ISSUE THREAD

**Issue by sottar** [styled-components/styled-components#1198]
How can I resolve this warning?

```
Warning: Received `true` for non-boolean attribute `gray`. If this is expected, cast the value to a string. warning.js:33 
```

The code is looks like this.
```
export const Button = styled(Link).attrs({
  width: props => props.width || '100%',
})`
  display: block;
  width: ${props => props.width};
  text-align: center;
  color: #fff;
  background-color: ${props => props.gray ? '#bababa' : '#ea352d'};
`;

export default class Hoge extends React.PureComponent {
  render() {
    return (
      <Button gray to='path/to/page'>hoge</Button>
    );
  }
}
```

When I passed ```gray``` props, the warning happened.

modules version.
- styled-components: 2.2.0
- react-router: 4.2.0
- react-router-dom: 4.2.2

**Comment by MisterTaki** [comment]
Temporarily:
    <Button gray={gray ? 1 : 0} to='path/to/page'>hoge</Button>

It works for me :)

**Comment by MisterTaki** [comment]
Temporarily:
    `<Button gray={gray ? 1 : 0} to='path/to/page'>hoge</Button>`

It works for me :)

**Comment by mxstbr** [comment]
When you style a standard React component with `styled(Comp)`, we pass through all the props you pass to the styled component:

```JS
const Comp = (props) => {
  // Note: Uses props.gray
  return <div>{props.gray}</div>
}

const StyledComp = styled(Comp)`
  color: blue;
`;

<StyledComp gray="Hello world" />
// -> This will render <div>Hello World</div>
// because the props you pass to <StyledComp /> are passed through to the underlying <Comp />
```

The `react-router` Link component passes all props you pass in to the underlying DOM node like so: `<a {...props />`. This means you end up with the `gray` property being attached to the DOM node as an attribute:

```JS
// This is what react-router <Link /> does under the hood:
const Link = (props) => {
  // it passes all props to the DOM node
  return <a {...props}>
}

// That means when you wrap it in styled(), all the props you pass to StyledLink are passed through
const StyledLink = styled(Link)`
  color: blue;
`;

<StyledLink gray={true} />
// -> This will render <a gray="true" />
// because the gray prop is passed through to <Link /> which passes it through to the DOM node
```

There is nothing we can do from the `styled-components` side of things to avoid this error. You'll have to either work around this by introducing a filtering component (see #439 and the discussion there) or by not passing any non-standard attributes to that component.

We should probably add a FAQ to the docs about this, cc @morajabi @philpl.

**Comment by cbroberg** [comment]
I get the same error with this code:

```js
const rotateLogo = keyframes`
    from {
        transform: rotate(0deg);
    }
	
	to {
        transform: rotate(360deg);
	}
`

const LogoImage = styled.img`
	  animation: ${(props) => props.rotate && `${rotateLogo} infinite 7s linear`};     
	  height: ${(props) => props.size || '100px'};
	  ${console.log((props) => props.rotate)}
`


export const Logo = (props) => {
	return (
		<LogoImage 
			src={logo} 
			size={props.size} 
			rotate={props.rotate} 
			alt={'ODEUM Code logo'} />
	)
}

Logo.propTypes = {
	rotate: PropTypes.bool,
	size: PropTypes.string,
	src: PropTypes.string,
	alt: PropTypes.string
}
```

**Comment by good-idea** [comment]
@sottar if you don't mind a little more nesting, you can do something like this:

```js
export const Button = styled(H4.withComponent('button'))`
	cursor: pointer;
	text-transform: uppercase;
 	padding: 10px;
 	display: inline-block;
 	min-width: 12rem;
 	text-align: center;

	${props => (props.secondary ?
		`border: none;
		color: gray;` :
		standardBorder
	)}
`

export const LinkButton = props => (
	<Button {...props}>
		<Link {...props}>{props.children}</Link>
	</Button>
)
```

**Comment by lnpbk** [comment]
I'm pretty new to both React, and styled-components, but I made myself a little helper for this problem

```javascript
const e = React.createElement;
const styleModifiers = ['primary']; // An array of prop names that are only used to modify the style
const resolveTagFromProps = (styleModifiers) => (({ tag, children, ...props }) => {
    styleModifiers.map(propName => !!props[propName] && delete props[propName]);
    return e(tag, props, children);
});

const StyledButton = styled(resolveTagFromProps(styleModifiers))`
    background: ${props => props.primary ? 'red' : 'blue'}
`

const Button = (props) => (
    <StyledButton {...props}>
        {props.children}
    </StyledButton>
);
```

Now I can safely use `<Button primary tag="a">Submit</Button>`

**Comment by schettino** [comment]
I think this way is a little bit cleaner

```javascript
const Pad = 'Pad';

const StyledGrid = styled(GridMaterial)`
  .${Pad} {
    @media (min-width: 768px) {
      padding: 24px;
    }
  }
`;

const Grid = ({ pad, ...props }) => (
  <StyledGrid  {...props} className={pad? Pad : undefined} />
)
```

**Comment by probablyup** [comment]
There's nothing to be done here. Just don't pass the prop to the DOM element if you're using a custom component.

**Comment by nmaves** [comment]
This seems to work for me.

```javascript
import React from 'react'

import { Link as RRLink } from 'react-router-dom'
import styled, { css } from 'styled-components'

import { themeValue } from 'theme'

const Link = ({ className, to, children }) => (
  <RRLink className={className} to={to}>
    {children}
  </RRLink>
)

export default styled(Link)`
  color: ${themeValue('colors.black')};
  :visited {
    color: ${themeValue('colors.black')};
  }
  text-decoration: none;

    /* PRIMARY */
    ${props =>
      props.primary &&
      css`
        color: ${themeValue('colors.primary.base')};

        &:visited {
          color: ${themeValue('colors.primary.base')};
        }
      `}

  /* ACCENT */
  ${props =>
    props.accent &&
    css`
      color: ${themeValue('colors.accent.base')};

      &:hover {
        color: ${themeValue('colors.accent.base')};
      }
    `}
`
```

**Comment by immortalObject** [comment]
You should filter all non-standard props:
```JSX
const StyledLink= styled(({ primary, ...props }) => <Link {...props} />)`
  color: ${props => props.primary ? themeColors.primary : 'red'};
`
```

**Comment by probablyup** [comment]
Worth noting we are looking into a [formal solution](https://github.com/styled-components/styled-components/pull/1682) for this (likely to land in v5).

**Comment by thecneu** [comment]
What are the whitelisted props that get passed through to base dom elements. I am trying to do `<Button fill />` and I get the non-boolean warning. I am guessing `fill` is being passed through due to it being an svg attribute. Is there anyway to make the `fill` prop work or should I find a new word to use?

**Comment by Sebastianpalm91** [comment]
Maybe a little bit too late, but this worked for me:

`<TabButton onClick={this.handleToggle} isActive={this.state.open}>{tab[this.state.tabIndex]}</TabButton>`

Sending this to the styled component as a prop:
> isActive={this.state.open}

and: 
`transform: ${props => props.isActive ? 'scaleX(0)' : 'scaleX(1)'};`

**Comment by axalix** [comment]
There are some special attributes that are naturally boolean, e.g. `[open](https://www.w3.org/TR/html51/interactive-elements.html#element-attrdef-details-open)` 
So you can always safely use them to keep the logic simple.

**Comment by marlonmleite** [comment]
> You should filter all non-standard props:
> 
> ```js-jsx
> const StyledLink= styled(({ primary, ...props }) => <Link {...props} />)`
>   color: ${props => props.primary ? themeColors.primary : 'red'};
> `
> ```

This work, but this is ugly and "painful".

**Comment by smakosh** [comment]
> Temporarily:
> `<Button gray={gray ? 1 : 0} to='path/to/page'>hoge</Button>`
> 
> It works for me :)

If this doesn't apply the style for you, try `null` instead of `0`

**Comment by linonetwo** [comment]
Hey, why not use `data-xxx`?

```jsx
export const Button = styled(Link)<{ 'data-visited': boolean }>`
  background-color: ${({ 'data-visited': visited }) =>
    visited &&
    css`
      background: gray;
    `};
`;

export default class Hoge extends React.PureComponent {
  render() {
	const isVisited = false;
    return (
      <Button data-visited={isVisited} to='path/to/page'>hoge</Button>
    );
  }
}
```

**Comment by somahargitai** [comment]
This solution in docs seems to make sense:
https://www.styled-components.com/docs/faqs#why-am-i-getting-html-attribute-warnings

**Comment by willredington** [comment]
I don't believe this issue is resolved when using typescript.  
I have the following scenario:

```tsx

// styles.tsx
const setOffset = ({ offset }: ListProps) => {
	if (offset) {
		return css`
			${ListItem} {
				margin-left: 0.15em;
			}
		`;
	}
};

export const List = styled.ul<ListProps>`
	padding-left: 5px;
	list-style-type: none;
	${setOffset};
`;

// props.tsx

export type ListProps = {
	offset?: boolean | 1 | 0 | "true" | "false";
};


// throws the error here
<List offset>{items}</List>

// no error here
<List offset="true">{items}</List>
```

**Comment by Hugofv** [comment]
``` javascript
import React from 'react';
import { Button } from "@material-ui/core";
import styled from "styled-components";

export default styled(({ primary, margintop, ...other }) => <Button {...other} />)`
  background-color: ${props => (props.primary ? "#CC131D !important" : '#EAEAEA !important')};
  color: ${props => ((props.primary && !props.disabled && "#FFF !important") || (!props.primary && !props.disabled && "#CC131D !important") || (props.disabled && "#6C6C6C !important"))};
  outline: none !important;
`;
```

**Comment by whazap** [comment]
i had a similar issue, but with the "as" prop
`<MyComponent as={Link} to="/my-url" customProp={customProp} />`

@immortalObject solution worked great
> You should filter all non-standard props:
> 
> ```js-jsx
> const StyledLink= styled(({ primary, ...props }) => <Link {...props} />)`
>   color: ${props => props.primary ? themeColors.primary : 'red'};
> `
> ```

slightly modified:
```js-jsx
<MyComponent
  as={({ to, className, children }) => (
    <Link to={to} className={className}>
      {children}
    </Link>
  )}
  to="/my-url"
  customProp={customProp}
/>
```

**Comment by luismramirezr** [comment]
> Material-UI does the same thing of React-Router.
> I solved by passing an object instead of a single prop:
> 
> ```js
> // card.js
> <StyledCardContent styled={{hasHeader: true}}>....</StyledCardContent>
> ```
> 
> and then:
> 
> ```js
> // card.style.js
> const StyledCardContent = styled(CardContent)` && {
>   ${({styled: {hasHeader}}) => hasHeader && `
>     padding-top: 0;
>   `}
> }`
> ```

This works!

**Comment by mrkvn** [comment]
This worked for me. I input any string as value of _gray_ and the warning did go away. In my case, "true" made sense.

`<Button gray="true" to='path/to/page'>hoge</Button>`

**Comment by georgemarin** [comment]
if helps anyboddy, I've used the following: 

`const EllipsisCell = styled(({ showDetails, ...props }) => <SemanticTable.Cell {...props} />)`
  overflow: visible !important;
  border-top: ${({ showDetails }) => showDetails && borderStyle};
`;
`

This way I can provide a boolean (instead of always using ternary like showDetails ? 1 : 0) and I can provide camel case props

**Comment by fandy** [comment]
The workaround for this error is to use an object as the value for the prop. For example:

```
<Component 			props={{
				transparent,
				scrolled,
			}} />

const Component = styled<any>(AppBar)(({ props }) => {
	return {
		display: 'flex',
		alignItems: 'center',
		margin: '0 auto',
		width: '100%',
		zIndex: 1,

		'&:before': {
			display: props.transparent && !props.scrolled ? 'block' : 'none',
			backgroundImage: 'linear-gradient(rgb(0, 0, 0), rgba(0, 0, 0, 0))',
			content: '""',
			height: '140%',
			left: 0,
			pointerEvents: 'none' as any,
			position: 'absolute' as any,
			top: 0,
			width: '100%',
			zIndex: 0,
			opacity: 0.4,
			transition: 'opacity 0.3s cubic-bezier(0.35, 0, 0.65, 1) 0.3s',
		},
	}
})

```

Wish there was a better workaround for this though.

**Comment by noahub** [comment]
I realize this is closed now, but I was having a similar issue and the fix for me was to use a transient prop:

https://styled-components.com/docs/api#transient-props

In my case it just involved changing my prop `isActive` to `$isActive`.

**Comment by babak-bekhrad** [comment]
Fixing this error is very simple. You just need to add + before your booleans values. According to MDN unary operator is the preferred way for number conversion. like this example:

```
<NavLink exact active={+true} to="/">
                  Home
 </NavLink>
```

**Comment by fbzz** [comment]
Maybe for someone who came later, take a look at:

https://styled-components.com/releases#:~:text=Add-,%22,-transient%20props%22%20API

**Comment by yracnet** [comment]
I fix this case with noInherit Attribute
```
import {  Card as BSCard } from 'react-bootstrap';
import styled from 'styled-components';

const BSCardHeader = ({ noInherit = [], ...props }) => {
  noInherit.forEach((it) => {
    delete props[it];
  });
  return <BSCard.Header {...props} />;
};

Card.Headline = styled(BSCardHeader)`
  ${(p) => p.slim && 'padding-top: 3px;padding-bottom: 3px;'};
  ${(p) => p.bold && 'font-weight: bold;'};
  ${(p) => p.upper && 'text-transform: uppercase;'};
`;

Card.Headline.defaultProps = {
  bg: 'light',
  color: 'black',
  slim: true,
  bold: true,
  upper: false,
  noInherit: ['slim', 'bold', 'upper'],
};
```

**Comment by MrVibe** [comment]
I am getting the warning even when the attribute is not being passed to the DOM.  Here's the warning I get : 

```
"client.js:26 Warning: Received `false` for a non-boolean attribute `order`.
If you want to write it to the DOM, pass a string instead: order="false" or order={value.toString()}.
If you used to conditionally omit it with order={condition && value}, pass order={condition ? value : undefined} instead."
```

I am using classnames which requires true or false for the classes :
```
<Tooltip message={t("shared.sortAscending")}>
    <ChevronUpIcon
        className={cx({"p-1 h-6 w-6 rounded": true, "bg-primary text-white": args.order})}
        onClick={() => setArgs({...args, order: true})}
    />
</Tooltip>
<Tooltip message={t("shared.sortDescending")}>
    <ChevronDownIcon
        className={cx({"p-1 h-6 w-6 rounded": true, "bg-primary text-white": !args.order})}
        onClick={() => setArgs({...args, order: false})}
    />
</Tooltip>
``` 
args.order is a boolean value. I get the error if I replace the args.error with true or false. Is there something obvious missing here ?
